apisix/plugins/authz-keycloak.lua

```diff
@@ -64,10 +67,169 @@ local _M = {
     schema = schema,
 }
 
+
 function _M.check_schema(conf)
     return core.schema.check(schema, conf)
 end
 
+
+-- Some auxiliary functions below heavily inspired by the excellent
+-- lua-resty-openidc module; see https://github.com/zmartzone/lua-resty-openidc
+
+
+-- Retrieve value from server-wide cache, if available.
+local function authz_keycloak_cache_get(type, key)
+  local dict = ngx.shared[type]
+  local value
+  if dict then
+    value = dict:get(key)
+    if value then log.debug("cache hit: type=", type, " key=", key) end
+  end
+  return value
+end
+
+
+-- Set value in server-wide cache, if available.
+local function authz_keycloak_cache_set(type, key, value, exp)
+  local dict = ngx.shared[type]
+  if dict and (exp > 0) then
+    local success, err, forcible = dict:set(key, value, exp)
+    if err then
+        log.error("cache set: success=", success, " err=", err, " forcible=", forcible)
+    else
+        log.debug("cache set: success=", success, " err=", err, " forcible=", forcible)
+    end
+  end
+end
+
+
+-- Configure timeouts.
+local function authz_keycloak_configure_timeouts(httpc, timeout)
+  if timeout then
+    if type(timeout) == "table" then
+      httpc:set_timeouts(timeout.connect or 0, timeout.send or 0, timeout.read or 0)
+    else
+      httpc:set_timeout(timeout)
+    end
+  end
+end
+
+
+-- Set outgoing proxy options.
+local function authz_keycloak_configure_proxy(httpc, proxy_opts)
+  if httpc and proxy_opts and type(proxy_opts) == "table" then
+    log.debug("authz_keycloak_configure_proxy : use http proxy")
+    httpc:set_proxy_options(proxy_opts)
+  else
+    log.debug("authz_keycloak_configure_proxy : don't use http proxy")
+  end
+end
+
+
+-- Parse the JSON result from a call to the OP.
+local function authz_keycloak_parse_json_response(response, ignore_body_on_success)
```

Comment:
What is the `ignore_body_on_success` for? It seems there is nowhere to pass it.

Reply:
I've taken some of these helper functions largely from the `openidc` module; see https://github.com/zmartzone/lua-resty-openidc

Agree that we don't need that parameter here. Will remove it.